File: app07_rust_react/backend/src/handlers/frameworks.rs
use axum::{
    Json,
    extract::{Path, State},
};

use crate::{
    error::{AppError, AppResult},
    models::Framework,
    state::AppState,
};

// Runtime-checked (`query_as`, not the `query_as!` macro) rather than
// compile-time-checked, unlike PLAN.md's original commitment - compile-time
// checking needs a live, already-migrated database present at `cargo build`
// time, which makes the build non-portable (fails for anyone who clones this
// repo without a running, pre-migrated Postgres). See ../CLAUDE.md.
pub async fn list_frameworks(State(state): State<AppState>) -> AppResult<Json<Vec<Framework>>> {
    let frameworks = sqlx::query_as::<_, Framework>(
        r#"SELECT id, code, name, version, description, reference_url
           FROM framework
           ORDER BY code"#,
    )
    .fetch_all(&state.pool)
    .await?;

    Ok(Json(frameworks))
}

pub async fn get_framework(
    State(state): State<AppState>,
    Path(code): Path<String>,
) -> AppResult<Json<Framework>> {
    let framework = sqlx::query_as::<_, Framework>(
        r#"SELECT id, code, name, version, description, reference_url
           FROM framework
           WHERE UPPER(code) = UPPER($1)"#,
    )
    .bind(&code)
    .fetch_optional(&state.pool)
    .await?
    .ok_or_else(|| AppError::NotFound("Framework", code.clone()))?;

    Ok(Json(framework))
}

File: app14_LUA_UNITY/backend/app/routes/frameworks.lua
local Framework = require("app.models.framework")

local function framework_entity(f)
  return {
    code = f.code,
    name = f.name,
    version = f.version,
    description = f.description,
    referenceUrl = f.reference_url
  }
end

return function(app)
  app:get("/api/v1/frameworks", function(self)
    local frameworks = Framework:select()
    local content = {}
    for _, f in ipairs(frameworks) do table.insert(content, framework_entity(f)) end
    return { json = content }
  end)

  app:get("/api/v1/frameworks/:code", function(self)
    local framework = Framework:find(self.params.code)
    if not framework then
      return {
        status = 404,
        json = {
          timestamp = os.date("!%Y-%m-%dT%H:%M:%SZ"),
          status = 404,
          error = "Not Found",
          message = "No such framework"
        }
      }
    end
    return { json = framework_entity(framework) }
  end)
end
